data/transactions/verify/txn.go

```diff
@@ -448,3 +456,394 @@ func (w *worksetBuilder) next() (txnGroups [][]transactions.SignedTxn) {
+		return false
+	}
+}
+
+func (sv *StreamVerifier) addVerificationTaskToThePool(uel unverifiedElementList, returnSeat bool) error {
+	sv.pendingTasksWg.Add(1)
+	function := func(arg interface{}) interface{} {
+		defer sv.pendingTasksWg.Done()
+
+		if sv.ctx.Err() != nil {
+			return struct{}{}
+		}
+
+		uel := arg.(*unverifiedElementList)
+		batchVerifier := crypto.MakeBatchVerifier()
+
+		bl := makeBatchLoad()
+		// TODO: separate operations here, and get the sig verification inside the LogicSig to the batch here
+		for _, ue := range uel.elementList {
+			groupCtx, err := txnGroupBatchPrep(ue.TxnGroup, uel.nbw.getBlockHeader(), uel.ledger, batchVerifier)
+			if err != nil {
+				// verification failed, no need to add the sig to the batch, report the error
+				sv.sendResult(ue.TxnGroup, ue.Context, err)
+				continue
+			}
+			totalBatchCount := batchVerifier.GetNumberOfEnqueuedSignatures()
+			bl.addLoad(ue.TxnGroup, groupCtx, ue.Context, totalBatchCount)
+		}
+
+		failed, err := batchVerifier.VerifyWithFeedback()
+		// this error can only be crypto.ErrBatchHasFailedSigs
+
+		verifiedTxnGroups := make([][]transactions.SignedTxn, 0, len(bl.txnGroups))
+		verifiedGroupCtxs := make([]*GroupContext, 0, len(bl.groupCtxs))
+		failedSigIdx := 0
+		for txgIdx := range bl.txnGroups {
+			txGroupSigFailed := false
+			// if err == nil, then all sigs are verified, no need to check for the failed
+			for err != nil && failedSigIdx < bl.messagesForTxn[txgIdx] {
+				if failed[failedSigIdx] {
+					// if there is a failed sig check, then no need to check the rest of the
+					// sigs for this txnGroup
+					failedSigIdx = bl.messagesForTxn[txgIdx]
+					txGroupSigFailed = true
+				} else {
+					// proceed to check the next sig belonging to this txnGroup
+					failedSigIdx++
+				}
+			}
+			var result error
+			if !txGroupSigFailed {
+				verifiedTxnGroups = append(verifiedTxnGroups, bl.txnGroups[txgIdx])
+				verifiedGroupCtxs = append(verifiedGroupCtxs, bl.groupCtxs[txgIdx])
+			} else {
+				result = ErrInvalidSignature
+			}
+			sv.sendResult(bl.txnGroups[txgIdx], bl.elementContext[txgIdx], result)
+		}
+		// loading them all at once by locking the cache once
+		sv.cache.AddPayset(verifiedTxnGroups, verifiedGroupCtxs)
+		return struct{}{}
+	}
+
+	// if the task has an allocated seat, should release it
+	var retChan chan interface{}
+	if returnSeat {
+		retChan = sv.seatReturnChan
+	}
+	// EnqueueBacklog returns an error when the context is canceled
+	err := sv.verificationPool.EnqueueBacklog(sv.ctx, function, &uel, retChan)
+	return err
+}
+
+func getNumberOfBatchableSigsInGroup(stxs []transactions.SignedTxn) (batchSigs uint64, err error) {
+	batchSigs = 0
+	for _, stx := range stxs {
+		count, err := getNumberOfBatchableSigsInTxn(stx)
+		if err != nil {
+			return 0, err
+		}
+		batchSigs = batchSigs + count
+	}
+	return
+}
+
+func getNumberOfBatchableSigsInTxn(stx transactions.SignedTxn) (batchSigs uint64, err error) {
```

Comment:
This function is pretty easy to figure out, but it could be broken into two-ish sub functions for slightly easier reading. One for Validation (ensure there is 0 or 1 signatures only), one to check if it's a StateProof, and then the counting could be in this function or in its own.